File: typescript/roman-numbers-reverse/src/Conversion.ts
import _ = require('lodash')

export class Conversion {
    constructor(private readonly remainingInput: number, private readonly digits?: string[]) {
        this.digits = digits || []
    }

    getResult() {
        return this.digits.join('')
    }

    apply(rule: { arabic: number, roman: string }) {
        const howManyDigits = this.remainingInput / rule.arabic
        const digits = this.digits.concat(_.times(howManyDigits, () => rule.roman))
        const remainingInput = this.remainingInput % rule.arabic
        return new Conversion(remainingInput, digits)
    }
}

File: kotlin/roman-numbers-reverse/src/main/kotlin/Conversion.kt
class Conversion(private val remainingInput: Int, private val digits: Array<String> = emptyArray()) {
    val result get() = digits.joinToString("")

    fun apply(rule: Rule): Conversion {
        val howManyVs = remainingInput / rule.arabic
        val newDigits = digits + Array(howManyVs) { rule.roman }
        val newRemainingInput = remainingInput % rule.arabic
        return Conversion(newRemainingInput, newDigits)
    }
}